<a href="https://colab.research.google.com/github/DexterWilliamson/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [0]:
import pandas as pd

In [0]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

### Specific Columns hold specific kinds of data

In [0]:
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
df.dtypes

Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [10]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr'].head()

0    3
1    3
2    3
3    3
4    4
Name: BedroomAbvGr, dtype: int64

Some columns hold float values like the `LotFrontage` column.

In [11]:
# Look at the first ten rows of the `BedroomAbvGr` column.
df['LotFrontage'].head()

0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
Name: LotFrontage, dtype: float64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [12]:
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [14]:
# NaN is the most common value in this column. What is a NaN
df['LotFrontage'].value_counts(dropna=False).head()

NaN     259
60.0    143
70.0     70
80.0     69
50.0     57
Name: LotFrontage, dtype: int64

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [15]:
import numpy as np

type(np.NaN)
# What is the datatype of NaN?

float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [0]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets

small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']]


### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [21]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

df.SalePrice

# Using bracket syntax to make a new 'TotalSquareFootage' column

df['SalePrice']

small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

small_df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [24]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['PricePerSqFt'] = round(small_df['SalePrice'] / small_df['TotalSquareFootage'],2)

small_df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0         81.25  
1         71.91  
2         82.59  
3         56.61  
4         74.78

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [25]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-01-14 17:51:26--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [                <=> ]  21.72M   419KB/s    in 56s     

2020-01-14 17:52:22 (399 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22776791]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [26]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [27]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [28]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [40]:
# Read in the CSV

df = pd.read_csv('LoanStats_2018Q4.csv', header=1)

df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65          0.0         Jun-2006             0.0   
1         IL  19.14          0.0         Oct-1996             3.0   
2         NE  22.23          0.0         May-2007             1.0   
3         MI   9.44          0.0         Mar-2009             2.0   
4         VA  18.47          0.0         Dec-2004             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       6.0      0.0   
1                    39.0                     NaN      14.0      0.0   
2                     NaN                     NaN       8.0      0.0   
3                    73.0                     NaN       7.0      0.0   
4                    75.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0     4196.0      91.2%        8.0                   w       0.00   
1     8547.0      27.7%       28.0                   w   16807.30   
2    25754.0        48%       23.0                   w       0.00   
3    11434.0      33.3%       20.0                   w    1707.34   
4    13555.0      38.2%       22.0                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymn

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [0]:
# We can fix the header problem by using the 'skiprows' parameter



Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [41]:
# Sum null values by column and sort from least to greatest
pd.set_option('display.max_rows', 200)
df.isnull().sum().sort_values()

inq_fi                                             2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
max_bal_bc                                         2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
hardship_flag                                      2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

In [64]:
# Address the extra NaNs in each column by skipping the footer as well.

df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2)

df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.
  


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       6000         6000           6000.0   36 months   
1 NaN        NaN      24000        24000          24000.0   36 months   
2 NaN        NaN      30000        30000          30000.0   36 months   
3 NaN        NaN       8000         8000           8000.0   36 months   
4 NaN        NaN      22400        22400          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65            0         Jun-2006               0   
1         IL  19.14            0         Oct-1996               3   
2         NE  22.23            0         May-2007               1   
3         MI   9.44            0         Mar-2009               2   
4         VA  18.47            0         Dec-2004               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN         6        0   
1                    39.0                     NaN        14        0   
2                     NaN                     NaN         8        0   
3                    73.0                     NaN         7        0   
4                    75.0                     NaN        15        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       4196      91.2%          8                   w       0.00   
1       8547      27.7%         28                   w   16807.30   
2      25754        48%         23                   w       0.00   
3      11434      33.3%         20                   w    1707.34   
4      13555      38.2%         22                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymnt_d  

In [65]:
df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
hardship_flag                                      0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [66]:
df = df.drop(['url', 'desc', 'id', 'member_id'], axis=1)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [67]:
# Look at the first 10 values of the int_rate column
df['int_rate'].head(10)

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

In [68]:
# Look at a specific value from the int_rate column

df['int_rate'][3]

' 12.98%'

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [0]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

int_rate = ' 12.98%'

In [70]:
int_rate.strip()

'12.98%'

In [71]:
int_rate.strip('%')

' 12.98'

In [72]:
int_rate.strip().strip('%')

'12.98'

In [73]:
float(int_rate.strip().strip('%'))

12.98

In [0]:
# "Cast" the string value to a float

In [0]:
def int_rate_to_float(cell_contents):

  return float(cell_contents.strip().strip('%'))

In [0]:
new_list = []

for cell in df['int_rate']:

  new_list.append(int_rate_to_float(cell))

In [0]:
df['int_rate_cleaned_from_apply'] = df['int_rate'].apply(int_rate_to_float)

In [76]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

### Write a function to make our solution reusable!

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

In [0]:
# Test out our function by calling it on our example

In [0]:
# is the data type correct?

### Apply our solution to every cell in a column

In [0]:
# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [77]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(dropna=False)[:20]

NaN                   20942
Teacher                2090
Manager                1773
Registered Nurse        951
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Truck Driver            387
Operations Manager      387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [78]:
# How many different unique employment titles are there currently?

df['emp_title'].value_counts()

Teacher                                2090
Manager                                1773
Registered Nurse                        951
Driver                                  924
RN                                      726
                                       ... 
Manager of Patient Engagement             1
Nutrition Educator                        1
Manager - Manufacturing Engineering       1
Estimating Manager Residential            1
Political Analyst                         1
Name: emp_title, Length: 43889, dtype: int64

In [80]:
len(df['emp_title'].value_counts())

43889

In [81]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20942

What are some possible reasons as to why a person's employment title may have not been provided?

In [0]:
# Create some examples that represent the cases that we want to clean up
examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [85]:
# Write a function to clean up these use cases and increase uniformity.

def clean_title(job_title):

  if isinstance(job_title, str):
    return job_title.title().strip()
  else:
    return "Unknown"

for title in examples:
  print(clean_title(title))  

Owner
Supervisor
Project Manager
Unknown


In [86]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.
[clean_title(example) for example in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [0]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column
df['emp_title'] = df['emp_title'].apply(clean_title)

We can use the same code as we did earlier to see how much progress was made.


In [88]:
# Look at the top 20 employment titles
df['emp_title'].head(20)

0                              Manager
1                  Maintenance Manager
2                          Firefighter
3                           Controller
4              Paramedic/ Fire Fighter
5                           Consultant
6                           Supervisor
7                              Unknown
8                 Warehouse Associates
9              Director Of Reinsurance
10                     Math Instructor
11                  Nursing Supervisor
12                            Security
13                      Administrative
14                    Material Handler
15           Instructional Coordinator
16    Financial Relationship Associate
17                             Unknown
18                 Sale Representative
19                  Driver Coordinator
Name: emp_title, dtype: object

In [89]:
# How many different unique employment titles are there currently?

df['emp_title'].value_counts(ascending=False)[:20]

Unknown               20942
Teacher                2557
Manager                2395
Registered Nurse       1417
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          804
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   424
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [90]:
# How often is the employment_title null (NaN)?

df['emp_title'].isnull().sum()

0

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [91]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [0]:
df['issue_d'][:10]

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [96]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [97]:
df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [100]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [102]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128392    2018
128393    2018
128394    2018
128395    2018
128396    2018
Name: issue_d, Length: 128397, dtype: int64

Now the month.

In [103]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128392    10
128393    10
128394    10
128395    10
128396    10
Name: issue_d, Length: 128397, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [105]:
df['issue_y'] = df['issue_d'].dt.year

df['issue_m'] = df['issue_d'].dt.month

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3              Firefighter  10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     55000.0            Verified 2018-12-01  Charged Off          n   
1    100000.0        Not Verified 2018-12-01      Current          n   
2     90000.0            Verified 2018-12-01   Fully Paid          n   
3     95000.0     Source Verified 2018-12-01      Current          n   
4     56000.0     Source Verified 2018-12-01      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019        

In [106]:
# Look at the datatypes of our new columns

df.dtypes[-2:]

issue_y    int64
issue_m    int64
dtype: object

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [107]:
df['earliest_cr_line'].head()

0    Jun-2006
1    Oct-1996
2    May-2007
3    Mar-2009
4    Dec-2004
Name: earliest_cr_line, dtype: object

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [111]:
df['issue_d'] - df['earliest_cr_line']

0         4566 days
1         8096 days
2         4232 days
3         3562 days
4         5113 days
            ...    
128392   12114 days
128393   13241 days
128394    4505 days
128395    3652 days
128396    4413 days
Length: 128397, dtype: timedelta64[ns]

In [0]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

What's oldest credit history that was involved in Q4 2018? 

In [117]:
df['days_from_earliest_credit_to_issue'].describe()

count    128397.000000
mean       5859.918464
std        2886.617678
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.